Src/IronPython/Runtime/PythonEncoding.cs

```diff
@@ -235,123 +322,179 @@ public abstract class PythonEncoderFallback : EncoderFallback, ICloneable {
         }
 
         protected abstract class PythonEncoderFallbackBuffer : EncoderFallbackBuffer {
-            private readonly char _marker;
-            private readonly Queue<byte> _fallbackBytes;
-
-            private struct ByteCounter {
-                private int _bcmCounter; // used when only counting bytes
-                private int _encCounter; // used during actual encoding
+            private struct RestorableInt {
```

Comment:
Tempted to say that this could be a readonly struct. Maybe not worth the effort if it works...

Reply:
It could be an interesting exercise. I am curious how it would compare.

Reply:
I had come up with something along these lines:
```C#
private readonly struct RestorableInt {
    private readonly int _value;
    private readonly int _saved;

    private RestorableInt(int value, int saved) {
        _value = value;
        _saved = saved;
    }

    public int Value => _value;
    public int Saved => _saved;

    public static implicit operator int(RestorableInt dc) => dc._value;
    public static implicit operator RestorableInt(int dc) => new RestorableInt(dc, dc);

    /// <summary>Assignment preserving the save</summary>
    public static RestorableInt operator <<(RestorableInt dc, int value) => new RestorableInt(value, dc._saved);

    public static RestorableInt operator + (RestorableInt dc, int value) => new RestorableInt(dc._value + value, dc._saved);
}
```